File: l2/z2/GF1234577.java
public class GF1234577 {
    private long value;
    private static final long CHARACTERISTIC = 1234577;

    public GF1234577(long value) {
        this.value = mod(value);
    }

    private static long mod(long x) {
        return ((x % CHARACTERISTIC) + CHARACTERISTIC) % CHARACTERISTIC;
    }

    private static long multiplicativeInverse(long a) {
        long x0 = 0, x1 = 1, temp;
        long b = CHARACTERISTIC;

        while (b != 0) {
            long q = a / b;
            long r = a % b;
            a = b;
            b = r;

            temp = x0;
            x0 = x1 - q * x0;
            x1 = temp;
        }

        if (a != 1) {
            throw new ArithmeticException("Inverse does not exist.");
        } else {
            return mod(x1);
        }
    }

    public GF1234577 add(GF1234577 rhs) {
        return new GF1234577(mod(this.value + rhs.value));
    }

    public GF1234577 subtract(GF1234577 rhs) {
        return new GF1234577(mod(this.value - rhs.value));
    }

    public GF1234577 multiply(GF1234577 rhs) {
        return new GF1234577(mod(this.value * rhs.value));
    }

    public GF1234577 divide(GF1234577 rhs) {
        if (rhs.value == 0) {
            throw new ArithmeticException("Division by zero is not defined.");
        }
        long inverse = multiplicativeInverse(rhs.value);
        return this.multiply(new GF1234577(inverse));
    }

    public boolean equals(GF1234577 rhs) {
        return this.value == rhs.value;
    }

    public boolean notEquals(GF1234577 rhs) {
        return this.value != rhs.value;
    }

    public boolean lessThan(GF1234577 rhs) {
        return this.value < rhs.value;
    }

    public boolean lessThanOrEqualTo(GF1234577 rhs) {
        return this.value <= rhs.value;
    }

    public boolean greaterThan(GF1234577 rhs) {
        return this.value > rhs.value;
    }

    public boolean greaterThanOrEqualTo(GF1234577 rhs) {
        return this.value >= rhs.value;
    }

    @Override
    public String toString() {
        return Long.toString(value);
    }

    public static void main(String[] args) {
        GF1234577 a = new GF1234577(10);
        GF1234577 b = new GF1234577(20);
        System.out.println("a + b = " + a.add(b).toString());
        System.out.println("a - b = " + a.subtract(b).toString());
        System.out.println("a * b = " + a.multiply(b).toString());
        System.out.println("a / b = " + a.divide(b).toString());
        System.out.println("a == b = " + a.equals(b));
        System.out.println("a != b = " + a.notEquals(b));
        System.out.println("a < b = " + a.lessThan(b));
        System.out.println("a <= b = " + a.lessThanOrEqualTo(b));
        System.out.println("a > b = " + a.greaterThan(b));
        System.out.println("a >= b = " + a.greaterThanOrEqualTo(b));
        GF1234577 c = new GF1234577(5);
        GF1234577 d = new GF1234577(0);
        System.out.println("c / d = " + c.divide(d).toString());
    }
}

File: l2/z1/GF1234577.cpp
#include <iostream>
#include <stdexcept>

class GF1234577
{
private:
    long value;
    static const long CHARACTERISTIC = 1234577;

    static long mod(long x)
    {
        return ((x % CHARACTERISTIC) + CHARACTERISTIC) % CHARACTERISTIC;
    }

    static long multiplicativeInverse(long a)
    {
        long x0 = 0, x1 = 1, temp;
        long b = CHARACTERISTIC;

        while (b != 0)
        {
            long q = a / b;
            long r = a % b;

            a = b;
            b = r;

            temp = x0;
            x0 = x1 - q * x0;
            x1 = temp;
        }

        if (a != 1)
        {
            throw std::runtime_error("Inverse does not exist.");
        }

        return mod(x1);
    }

public:
    GF1234577(long val = 0) : value(mod(val)) {}

    GF1234577(const GF1234577 &other) : value(other.value) {}

    GF1234577 &operator=(const GF1234577 &other)
    {
        value = other.value;
        return *this;
    }

    GF1234577 operator+(const GF1234577 &rhs) const
    {
        return GF1234577(mod(value + rhs.value));
    }

    GF1234577 operator-(const GF1234577 &rhs) const
    {
        return GF1234577(mod(value - rhs.value));
    }

    GF1234577 operator*(const GF1234577 &rhs) const
    {
        return GF1234577(mod(value * rhs.value));
    }

    GF1234577 operator/(const GF1234577 &rhs) const
    {
        if (rhs.value == 0)
        {
            throw std::runtime_error("Division by zero is not defined.");
        }
        return (*this) * GF1234577(multiplicativeInverse(rhs.value));
    }

    GF1234577 &operator+=(const GF1234577 &rhs)
    {
        value = mod(value + rhs.value);
        return *this;
    }

    GF1234577 &operator-=(const GF1234577 &rhs)
    {
        value = mod(value - rhs.value);
        return *this;
    }

    GF1234577 &operator*=(const GF1234577 &rhs)
    {
        value = mod(value * rhs.value);
        return *this;
    }

    GF1234577 &operator/=(const GF1234577 &rhs)
    {
        (*this) = (*this) / rhs;
        return *this;
    }

    bool operator==(const GF1234577 &rhs) const
    {
        return value == rhs.value;
    }

    bool operator!=(const GF1234577 &rhs) const
    {
        return !(*this == rhs);
    }

    bool operator<(const GF1234577 &rhs) const
    {
        return value < rhs.value;
    }

    bool operator<=(const GF1234577 &rhs) const
    {
        return value <= rhs.value;
    }

    bool operator>(const GF1234577 &rhs) const
    {
        return value > rhs.value;
    }

    bool operator>=(const GF1234577 &rhs) const
    {
        return value >= rhs.value;
    }

    friend std::ostream &operator<<(std::ostream &os, const GF1234577 &obj)
    {
        os << obj.value;
        return os;
    }
};

int main()
{
    GF1234577 a(10), b(20);
    std::cout << "a + b = " << (a + b) << std::endl;
    std::cout << "a - b = " << (a - b) << std::endl;
    std::cout << "a * b = " << (a * b) << std::endl;
    std::cout << "a / b = " << (a / b) << std::endl;
    std::cout << "a == b = " << (a == b) << std::endl;
    std::cout << "a != b = " << (a != b) << std::endl;
    std::cout << "a < b = " << (a < b) << std::endl;
    std::cout << "a <= b = " << (a <= b) << std::endl;
    std::cout << "a > b = " << (a > b) << std::endl;
    std::cout << "a >= b = " << (a >= b) << std::endl;
    GF1234577 c(5), d(0);
    std::cout << "c / d = " << (c / d) << std::endl;

    return 0;
}